Add proof generation and verification in tests

phklive opened the pull request:
In this PR I propose an addition of proof generation and verification in tests. The miden-base repo has multiple crates that run multiple tests that create and execute transactions. Up to now these transactions were executed but not verified, hence preventing the correct verification of correct flow.

Now the tests generate a proof and verify the proof of correct execution of generated the transactions.

P2ID: Added proof and verification on tx
SWAP: Added proof and verification on tx
P2IDR: Added proof and verification only on first tx
Faucet: Added proof and verification only on `mint` and `burn`
Wallet: Added proof and verification on `send` and `receive` asset

Changed all `transaction_result` to `executed_transaction` for consistency.

This PR adds 7 proof generated and verified, total in tests: 8

Built on top of: #435 

Will need to wait to merge before merging this one: #439

bobbinth commented on miden-tx/tests/p2id_script_test.rs:
```diff
@@ -235,13 +236,83 @@ fn test_p2id_script_multiple_assets() {
+    let note_ids = data_store.notes.iter().map(|note| note.id()).collect::<Vec<_>>();
+
+    let tx_script_code = ProgramAst::parse(
+        "
+            use.miden::contracts::auth::basic->auth_tx
+    
+            begin
+                call.auth_tx::auth_tx_rpo_falcon512
+            end
+            ",
```
We use this string several time in this file and probably in others as well. Let's create a common constant for it.

phklive replied:
Added in mock constants: 

```rust
// Default authentication code
pub const DEFAULT_AUTH_CODE: &str = "
    use.miden::contracts::auth::basic->auth_tx

    begin
        call.auth_tx::auth_tx_rpo_falcon512
    end
";
```

bobbinth commented on miden-tx/tests/p2id_script_test.rs:
```diff
@@ -235,13 +236,83 @@ fn test_p2id_script_multiple_assets() {
+    let proven_transaction = prover.prove_transaction(executed_transaction.clone()).unwrap();
+
+    // Verify that the generated proof is valid
+    let verifier = TransactionVerifier::new(96);
```
Let's create a constant for `96` as well - e.g., `MIN_PROOF_SECURITY_LEVEL`.

phklive replied:
https://github.com/0xPolygonMiden/miden-base/pull/438#discussion_r1467481737

bobbinth commented on miden-tx/tests/swap_script_test.rs:
```diff
@@ -107,7 +108,82 @@ fn test_swap_script() {
     assert_eq!(created_note, &requested_note);
 }
+
+#[test]
+fn test_swap_prove_and_verify() {
```
Is there a benefit to adding a separate test like this? In `P2IDR` test, we just integrated proving/verification into the existing test. Should we do the same here?

phklive replied:
agreed.

bobbinth commented on miden-tx/tests/p2id_script_test.rs:
```diff
@@ -235,13 +236,83 @@ fn test_p2id_script_multiple_assets() {
+}
+
+#[test]
+fn test_p2id_prove_and_verify() {
```
Is there a benefit to adding a separate test like this? In `P2IDR` test, we just integrated proving/verification into the existing test. Should we do the same here?

phklive replied:
agreed.

bobbinth commented on miden-tx/tests/wallet_test.rs:
```diff
@@ -163,10 +174,20 @@ fn test_send_asset_via_wallet() {
+    let proven_transaction = prover.prove_transaction(executed_transaction.clone()).unwrap();
+
+    // Verify that the generated proof is valid
+    let verifier = TransactionVerifier::new(96);
```
Similar comment to one of the previous comments: let's use a constant for 96.

bobbinth reviewed (changes requested):
Thank you! Looks good! I left some comments inline - but they are pretty small.

Also, how long does this take to run? (I guess we can get the CI runtime on the next commit). I'm curious if 7 such tests are fine, or if we should reduce to 3.

phklive commented:
@bobbinth, In your comment you said that you believe that we could reduce the prove / verify to 3 tests. To be able to implement that I would need to understand: 
- What are the core logics where we want to make sure that prove / verify works for? ( Example: Do we want to verify all note scripts (P2ID, R, SWAP) or are we okay to test only one )
- Do you have some tests in mind that are essentials to prove and verify
- Are there tests that we can skip, they should be fine without

With this knowledge I should be able to make an informed choice on which I could skip and which I should prove / verify.

bobbinth commented:
> In your comment you said that you believe that we could reduce the prove / verify to 3 tests.

Let's see how long tests take first. Could you rebase your PR from the latest `main` - this way we'll get the correct numbers.

> What are the core logics where we want to make sure that prove / verify works for? ( Example: Do we want to verify all note scripts (P2ID, R, SWAP) or are we okay to test only one )

I think all of these tests roughly test the same thing - i.e., transaction prologue/epilogue for relatively simple transactions. So, picking one of them would probably be sufficient for now.

>  Do you have some tests in mind that are essentials to prove and verify

I think would be good to cover the following cases:

1. A test of a transaction which consumes a single note.
2. A test of a transaction which produces a single note.
3. A test of a transaction which consumes a note and produces a note.
4. A test of a transaction which consumes more than 1 note and produces more than 1 note.

Points 1 and 2 we can probably cover with current tests. For 3 and 4 we should create a separate issue.

> Are there tests that we can skip, they should be fine without

I think as far as executing transactions goes we shouldn't skip anything. But for proving and verifying, we'll be in a good shape if we have the 4 above cases covered.

bobbinth commented on miden-tx/tests/wallet_test.rs:
```diff
@@ -66,29 +67,28 @@ fn test_receive_asset_via_wallet() {
         .unwrap();
 
+    // Prove the transaction
+    let proof_options = ProvingOptions::default();
+    let prover = TransactionProver::new(proof_options);
+    let proven_transaction = prover.prove_transaction(executed_transaction.clone()).unwrap();
+
+    // Verify that the generated proof is valid
+    let verifier = TransactionVerifier::new(MIN_PROOF_SECURITY_LEVEL);
```
As a part of these tests, let's also serialize and then deserialize `ProvenTransaction` to simulate it going over the wire.

phklive replied:
> As a part of these tests, 

Do you mean in these two tests: 

- `test_receive_asset_via_wallet`: 
https://github.com/0xPolygonMiden/miden-base/blob/d885a6e5343fde2aef33623d41ad6aadccce9fc9/miden-tx/tests/wallet_test.rs#L25-L107

- `test_send_asset_via_wallet`:
https://github.com/0xPolygonMiden/miden-base/blob/d885a6e5343fde2aef33623d41ad6aadccce9fc9/miden-tx/tests/wallet_test.rs#L109-L195

Or do you mean in all tests that prove and verify txs?

bobbinth replied:
> Or do you mean in all tests that prove and verify txs?

I think whichever tests we end up doing prove-and-very in, we should serialize `ProvenTransaction`'s and then deserialize them before execution (performance overhead of this serialization/deserialization should be negligible).

phklive commented:
> I think would be good to cover the following cases:

For case 3 I think this would be a good candidate to test the new `SWAP` api. It would be in reverse where we would create a note and then consume a note. 
1. create swap script function is called
2. swap script is created, returns a P2ID note
3. swap script note is consumed, creates a P2ID note
3. P2ID note is consumed

What do you think? 

> Points 1 and 2 we can probably cover with current tests. For 3 and 4 we should create a separate issue.

Issue has been created here: #446

> I think as far as executing transactions goes we shouldn't skip anything.

When I said `skip` I meant, tests where we should `skip` putting in tx proving and verifying because it was not useful and could slow us down. 

> But for proving and verifying, we'll be in a good shape if we have the 4 above cases covered.

Perfect!

bobbinth commented:
> For case 3 I think this would be a good candidate to test the new `SWAP` api. It would be in reverse where we would create a note and then consume a note.

Actually, testing full flow of the `SWAP` could test all 3 cases:

1. Transaction 1: produces SWAP note.
2. Transaction 2: consumes SWAP note and produces payback note.
3. Transaction 3: consumes payback note.

phklive commented:
> > For case 3 I think this would be a good candidate to test the new `SWAP` api. It would be in reverse where we would create a note and then consume a note.
> 
> Actually, testing full flow of the `SWAP` could test all 3 cases:
> 
> 1. Transaction 1: produces SWAP note.
> 2. Transaction 2: consumes SWAP note and produces payback note.
> 3. Transaction 3: consumes payback note.

Have we been able to go forward on the Note refactor that you had in mind to enable the returning of notes from SWAP?

bobbinth commented:
> Have we been able to go forward on the Note refactor that you had in mind to enable the returning of notes from SWAP?

Still on my TODO list. Hopefully, will wrap it up by EOW.

phklive commented:
> > For case 3 I think this would be a good candidate to test the new `SWAP` api. It would be in reverse where we would create a note and then consume a note.
> 
> Actually, testing full flow of the `SWAP` could test all 3 cases:
> 
> 1. Transaction 1: produces SWAP note.
> 2. Transaction 2: consumes SWAP note and produces payback note.
> 3. Transaction 3: consumes payback note.

Should we then open an issue to refactor the test of the SWAP note?

phklive commented:
For now the tests take approx 10 min: 

```bash
make test  912.81s user 25.70s system 156% cpu 9:59.62 total
```

bobbinth commented on mock/src/constants.rs:
```diff
@@ -39,6 +41,15 @@ pub const DEFAULT_ACCOUNT_CODE: &str = "
 ";
 
+// Default authentication code
+pub const DEFAULT_AUTH_CODE: &str = "
```
nit: I would change this to `DEFAULT_AUTH_SCRIPT`.

phklive replied:
Should this be called `DEFAULT_ACCOUNT_SCRIPT` then?

```rust
// Default account code
pub const DEFAULT_ACCOUNT_CODE: &str = "
    use.miden::contracts::wallets::basic->basic_wallet
    use.miden::contracts::auth::basic->basic_eoa

    export.basic_wallet::receive_asset
    export.basic_wallet::send_asset
    export.basic_eoa::auth_tx_rpo_falcon512
";
```

bobbinth replied:
We actually use the term "script" when we talk about notes and term "code" when we talk about accounts. The distinction is that "script" is an executable program while account code is a collection of procedures.

So, keeping this as `DEFAULT_ACCOUNT_CODE` is fine.

bobbinth commented on miden-tx/tests/wallet_test.rs:
```diff
@@ -66,29 +69,32 @@ fn test_receive_asset_via_wallet() {
         .execute_transaction(target_account.id(), block_ref, &note_ids, Some(tx_script))
         .unwrap();
 
+    // Prove the transaction
+    let proof_options = ProvingOptions::default();
+    let prover = TransactionProver::new(proof_options);
+    let proven_transaction = prover.prove_transaction(executed_transaction.clone()).unwrap();
+
+    // Serialize & deserialize the ProvenTransaction
+    let serialised_transaction = proven_transaction.to_bytes();
+    let proven_transaction = ProvenTransaction::read_from_bytes(&serialised_transaction).unwrap();
+
+    // Verify that the generated proof is valid
+    let verifier = TransactionVerifier::new(MIN_PROOF_SECURITY_LEVEL);
+
+    assert!(verifier.verify(proven_transaction).is_ok());
```
Should we encapsulate this in a function? Seems like this is being repeated in various tests. Could be something like `prove_and_verify_transaction()`.

phklive replied:
Added this function into `common`, only place where this function is note usable is in `tests.rs`, This will be fixed when we merge the test structure refactor: #445

bobbinth reviewed (changes requested):
Thank you! Looks good! I left a couple of comments inline.

> Takes 21 min in Github CI

The testing time seems way too high. We should try to get it under 1 min on an M1 and 1 - 5 mins in the CI.
- It is not clear why the tests take this much. My understanding is that under regular conditions, we can prove a transaction in about 5 seconds on an M1 - so, even with 9 tests, it shouldn't take more than 1 minute.
- One thing we discussed before was reducing the number of prove-and-verify tests to 3. If after we figure out what's going on above, we are still over 5 mins in the CI, we should reduce the number of prove-and-verify checks (we can still keep the execution tests, just remove proving and verifying part).

> Should we then open an issue to refactor the test of the SWAP note?

Yes, but let's do this after note refactoring is complete.

bobbinth commented:
After https://github.com/0xPolygonMiden/miden-vm/pull/1227 has been merged, the tests seem to be passing again.

A few other thoughts:
- I just remembered that each `.rs` file in the `src/tests` directory gets compiled into a separate crate. So, having many files in this directory blows up compilation time. The approach we've taken in Miden VM is to put integration tests `src/tests/integration` directory (see [here](https://github.com/0xPolygonMiden/miden-vm/tree/main/miden/tests/integration)). I don't remember all the detail, but I think it worked.
- I like the current approach of separating tests that involve proving from other tests. One small nit is that maybe instead of 'has_proving` suffix, we use `prove_` prefix for such tests?
- Lastly, we should document the testing approach in a readme (we don't have a readme for now at all, but let's create one with just "Testing" section, similar to how we have it [miden-crypto](https://github.com/0xPolygonMiden/crypto?tab=readme-ov-file#testing)).

phklive commented:
Working on improving the compilation times.

phklive commented:
Added the `integration` folder, indeed improved compilation time dropping from: 
- 10 > minutes: https://github.com/0xPolygonMiden/miden-base/actions/runs/7754493180 
- 5 < minutes: https://github.com/0xPolygonMiden/miden-base/actions/runs/7755108929

Will continue working on improving the compilation time in this PR refactoring the tests: 

I propose an incremental approach to the documentation of the miden-base considering:
1. It's a large and non-trivial repo, explanation for it is required
2. It is crucial for contributors to be able to understand the base being a central piece of Miden

By incremental I mean that each person contributing to the base should PR by PR add some documentation to the repo, Ex: Someone is working on note scripts, he should be in the best position to add the Note scripts section to the documentation of the repo. ( this is an example, I don't know for now what sections are essential in the base repo documentation, I believe that this is an interesting discussion that we could have )

bobbinth reviewed (changes requested):
Thank you! Not a full review - but I did notice that `prove_` tests don't actually run. We may need to do more here than just create `integration` directory (e.g., check out how Miden VM `Cargo.toml` is structured.

phklive commented:
Per your latest review I believe that I have fixed the problems that we're pointed out. I have removed `prove_` in the Makefile and have fixed the test file structure to make sure that all tests run correctly. I have checked my work for correctness as you requested yesterday during the call.

There are still some tests that run in CI (not on my computer) for more than 60 seconds: https://github.com/0xPolygonMiden/miden-base/actions/runs/7798779048/job/21268164800?pr=438#step:4:302
Let me know if we have a way to address that or if is okay and only related to the power of the GH CI machine.

bobbinth commented on README.md:
````diff
@@ -1 +1,15 @@
+make test
+```
+
+Some of the functions in this project are computationally intensive and may take a significant amount of time to compile and complete during testing. To ensure optimal results, it is recommended to run the tests in release mode, which using specific configurations replicates the test conditions of the development mode and verifies all debug assertions. For more information refer to the [Makefile] ...
````
nit: it is not super clear from this that this is what the `make test` command already does. Maybe we should say this explicitly?

phklive replied:
Made it more explicit.

bobbinth reviewed (approved):
Looks good! Thank you! I left some comments in the PR - but they are all pretty small. Once these are addressed, we can merge.

> There are still some tests that run in CI (not on my computer) for more than 60 seconds: https://github.com/0xPolygonMiden/miden-base/actions/runs/7798779048/job/21268164800?pr=438#step:4:302
> Let me know if we have a way to address that or if is okay and only related to the power of the GH CI machine.

I think it is fine as is and the overall CI times look pretty good. We could further consolidate proving tests as mentioned in https://github.com/0xPolygonMiden/miden-base/pull/438#issuecomment-1915993375 - but that's for a different PR (maybe let's create an issue for it though?)

phklive commented on miden-tx/tests/integration/scripts/faucet.rs:
```diff
@@ -15,16 +15,16 @@ use miden_objects::{
 
 #[test]
-fn test_faucet_contract_mint_fungible_asset_succeeds() {
+fn prove_test_faucet_contract_mint_fungible_asset_succeeds() {
```
I have only added `prove_` to the test names. the previous name started with `test_`, it also bothered me and I thought of changing it previously, if I have you approval I will remove the `test_` prefix

bobbinth replied:
Yes - for these tests, let's just go with `prove_` prefix.

phklive replied:
I have removed all of the `test_` prefixes on tests in miden-tx, they were all redundant imo.

phklive commented on Makefile:
```diff
@@ -1,8 +1,14 @@
+	cargo test $(PROFILE_TEST) $(FEATURES_CONCURRENT_TESTING) -- --skip prove
+	cargo test $(PROFILE_RELEASE) $(FEATURES_CONCURRENT_TESTING) prove
+
+pre:
+	cargo +nightly clippy
+	cargo +nightly fmt
```
Yes indeed this is the idea, have a single command before pushing new commits. 

Thank you for the detailed commands.

phklive commented:
> Looks good! Thank you! I left some comments in the PR - but they are all pretty small. Once these are addressed, we can merge.
> 
> > There are still some tests that run in CI (not on my computer) for more than 60 seconds: https://github.com/0xPolygonMiden/miden-base/actions/runs/7798779048/job/21268164800?pr=438#step:4:302
> > Let me know if we have a way to address that or if is okay and only related to the power of the GH CI machine.
> 
> I think it is fine as is and the overall CI times look pretty good. We could further consolidate proving tests as mentioned in [#438 (comment)](https://github.com/0xPolygonMiden/miden-base/pull/438#issuecomment-1915993375) - but that's for a different PR (maybe let's create an issue for it though?)

Could you be clearer on what you would need in this issue please? 

I believe that all comments have been addressed, should be ready for squash and merge.

phklive commented:
> I think it is fine as is and the overall CI times look pretty good. We could further consolidate proving tests as mentioned in https://github.com/0xPolygonMiden/miden-base/pull/438#issuecomment-1915993375 - but that's for a different PR (maybe let's create an issue for it though?)

Could you please clarify what you would want in this issue?

bobbinth commented:
> Could you please clarify what you would want in this issue?

Actually, I think #446 already covers this. So - nothing else is needed.
